File: apps/runtime/src/git/repository/integration/merge.rs
//! Starting a merge and resuming an owned integration sequence.

use super::*;

impl RepositoryService {
    pub(in crate::git::repository) async fn start_merge(
        &self,
        context: &RepositoryContext,
        target_oid: &str,
        message: &str,
        expected_token: &str,
        expected: &ExpectedState,
        operation: &Operation,
    ) -> AppResult<()> {
        let token = &operation.cancellation;
        let state = self.integration_snapshot(context, token).await?;
        if state.kind != "none" {
            return Err(AppError::Conflict(
                "An existing Git operation must be completed first".into(),
            ));
        }
        if state.head != *expected || state.state_token != expected_token {
            return Err(AppError::Conflict(
                "Repository state changed; refresh and confirm the merge again".into(),
            ));
        }
        if state.dirty
            || !state.conflicts.is_empty()
            || state.head.head_oid.is_none()
            || state.head.branch.is_none()
        {
            return Err(AppError::Conflict(
                "Merge requires a clean worktree and index on a committed local branch".into(),
            ));
        }
        if self.resolve(&context.repository, target_oid, token).await? != target_oid {
            return Err(AppError::BadRequest(
                "Merge target must be a commit object ID".into(),
            ));
        }
        let touched = self
            .read(
                &context.repository,
                args(&[
                    "diff",
                    "--no-ext-diff",
                    "--no-textconv",
                    "--name-only",
                    "--no-renames",
                    "-z",
                    expected.head_oid.as_deref().ok_or_else(malformed)?,
                    target_oid,
                    "--",
                ]),
                token,
            )
            .await?;
        let touched = touched
            .split(|byte| *byte == 0)
            .filter(|path| !path.is_empty())
            .map(|path| {
                std::str::from_utf8(path)
                    .map(str::to_owned)
                    .map_err(|_| malformed())
            })
            .collect::<AppResult<std::collections::BTreeSet<_>>>()?;
        self.protect_local_paths(context, &touched, token).await?;
        let session_id = operation.snapshot.lock().expect("Git operation").id.clone();
        self.inner
            .integrations
            .lock()
            .expect("Git integrations")
            .insert(
                context.repository.clone(),
                IntegrationOwner {
                    session_id,
                    kind: "merge",
                    mainline: None,
                    original: expected.clone(),
                    target_oid: target_oid.into(),
                    marker: None,
                },
            );
        let mut command = args(&[
            "-c",
            "rerere.enabled=false",
            "merge",
            "--no-ff",
            "--no-commit",
            "--no-autostash",
            "--no-overwrite-ignore",
            "--no-rerere-autoupdate",
        ]);
        if !message.is_empty() {
            command.extend(["-m".into(), message.into()]);
        }
        command.extend(["--".into(), target_oid.into()]);
        let result = self
            .output(
                &context.repository,
                command,
                self.command_timeout,
                token,
                Some(operation.mutation_started.clone()),
            )
            .await;
        // Use a new read token after a cancelled child has been reaped. This can
        // bind its actual remaining marker without ever retrying the mutation.
        let actual = self
            .git_integration(context, &Cancellation::default())
            .await;
        let actual_head = self
            .head(&context.repository, &Cancellation::default())
            .await;
        let mut owns = false;
        if let (Ok(actual), Ok(head)) = (&actual, &actual_head) {
            let mut owners = self.inner.integrations.lock().expect("Git integrations");
            if let Some(owner) = owners.get_mut(&context.repository) {
                if actual.kind == "merge"
                    && actual.target_oid.as_deref() == Some(target_oid)
                    && actual.original_head == expected.head_oid
                    && head == expected
                {
                    owner.marker = actual.marker.clone();
                    owns = owns_integration(owner, actual, head);
                } else if actual.kind == "none" {
                    owners.remove(&context.repository);
                }
            }
        }
        let output = result?;
        let actual = actual?;
        let actual_head = actual_head?;
        if owns && matches!(output.status, Some(0 | 1)) {
            operation.awaiting_resolution.store(true, Ordering::SeqCst);
            return Ok(());
        }
        if output.status != Some(0) {
            return Err(command_error(&output));
        }
        if actual.kind != "none" || actual_head != *expected {
            return Err(AppError::Conflict("Merge ownership or HEAD could not be verified after Git exited; inspect current state".into()));
        }
        let ancestor = self
            .output(
                &context.repository,
                args(&[
                    "merge-base",
                    "--is-ancestor",
                    target_oid,
                    expected.head_oid.as_deref().ok_or_else(malformed)?,
                ]),
                Duration::from_secs(15),
                token,
                None,
            )
            .await?;
        if ancestor.status != Some(0) {
            return Err(AppError::Conflict(
                "Git returned without a merge state or reachable target; inspect the repository"
                    .into(),
            ));
        }
        Ok(())
    }

    pub(in crate::git::repository) async fn resume_integration(
        &self,
        context: &RepositoryContext,
        session_id: &str,
        expected_token: &str,
        expected: &ExpectedState,
        recovery: Recovery,
        operation: &Operation,
    ) -> AppResult<()> {
        let token = &operation.cancellation;
        let state = self.integration_snapshot(context, token).await?;
        if !state.owned
            || state.session_id.as_deref() != Some(session_id)
            || state.head != *expected
            || state.state_token != expected_token
        {
            return Err(AppError::Conflict(
                "Integration ownership or repository state changed; refresh before continuing"
                    .into(),
            ));
        }
        if recovery == Recovery::Continue && !state.can_continue {
            return Err(AppError::Conflict(
                "Resolve all conflicts and stage the saved results before continuing".into(),
            ));
        }
        if recovery == Recovery::Skip && !state.can_skip {
            return Err(AppError::Conflict("Only an owned empty cherry-pick can be skipped; nonempty work must be resolved or explicitly aborted".into()));
        }
        if recovery != Recovery::Continue {
            self.protect_abort_paths(
                context,
                state.original_head.as_deref().ok_or_else(malformed)?,
                token,
            )
            .await?;
        }
        let command = match (state.kind.as_str(), recovery) {
            ("merge", Recovery::Abort) => args(&["merge", "--abort"]),
            ("merge", Recovery::Continue) => args(&["-c", "core.editor=:", "commit", "--no-edit"]),
            ("cherryPick", mode) => args(&[
                "-c",
                "core.editor=:",
                "cherry-pick",
                match mode {
                    Recovery::Continue => "--continue",
                    Recovery::Abort => "--abort",
                    Recovery::Skip => "--skip",
                },
            ]),
            // Skip belongs to an empty cherry-pick only; dropping a revert
            // silently would leave the change it was meant to undo in place.
            ("revert", mode @ (Recovery::Continue | Recovery::Abort)) => args(&[
                "-c",
                "core.editor=:",
                "revert",
                if mode == Recovery::Continue {
                    "--continue"
                } else {
                    "--abort"
                },
            ]),
            // Skip would silently drop a whole replayed commit, so a rebase
            // only offers the two decisions the caller can actually review.
            ("rebase", mode @ (Recovery::Continue | Recovery::Abort)) => args(&[
                "-c",
                "core.editor=:",
                "-c",
                "rerere.enabled=false",
                "rebase",
                if mode == Recovery::Continue {
                    "--continue"
                } else {
                    "--abort"
                },
            ]),
            _ => {
                return Err(AppError::Conflict(
                    "This Git operation is not managed by the current service".into(),
                ));
            }
        };
        self.mutate(context, command, operation).await?;
        let after = self.git_integration(context, token).await?;
        let head = self.head(&context.repository, token).await?;
        // Abort returns a rebase to its recorded branch, not to the detached
        // head the caller confirmed; every other kind leaves HEAD where it was.
        let restored = if state.kind == "rebase" {
            ExpectedState {
                head_oid: state.original_head.clone(),
                branch: state.original_branch.clone(),
            }
        } else {
            expected.clone()
        };
        if state.kind == "rebase" && recovery == Recovery::Continue && after.kind == "rebase" {
            // A replay can stop again on the next commit. That is still the
            // same owned sequence, not a failed recovery.
            let still_owned = self
                .inner
                .integrations
                .lock()
                .expect("Git integrations")
                .get(&context.repository)
                .is_some_and(|owner| owns_integration(owner, &after, &head));
            if still_owned {
                operation.awaiting_resolution.store(true, Ordering::SeqCst);
                return Ok(());
            }
        }
        if after.kind != "none" || (recovery != Recovery::Continue && head != restored) {
            return Err(AppError::Conflict("Git did not finish the requested recovery; inspect the current state before another action".into()));
        }
        if recovery == Recovery::Continue && state.kind == "cherryPick" {
            self.verify_cherry_pick(
                context,
                state.target_oid.as_deref().ok_or_else(malformed)?,
                expected,
                &head,
                token,
            )
            .await?;
        }
        if recovery == Recovery::Continue && state.kind == "revert" {
            self.verify_revert(
                context,
                state.target_oid.as_deref().ok_or_else(malformed)?,
                expected,
                &head,
                token,
            )
            .await?;
        }
        if recovery == Recovery::Continue && state.kind == "rebase" {
            self.verify_rebase(context, &state, &head, token).await?;
        }
        if recovery == Recovery::Continue && state.kind == "merge" {
            let commit = head.head_oid.as_deref().ok_or_else(malformed)?;
            let parents = self
                .read(
                    &context.repository,
                    args(&["rev-list", "--parents", "-n", "1", commit, "--"]),
                    token,
                )
                .await?;
            let parents = one_line(&parents)?
                .split_whitespace()
                .skip(1)
                .collect::<Vec<_>>();
            if head.branch != expected.branch
                || parents
                    != vec![
                        expected.head_oid.as_deref().ok_or_else(malformed)?,
                        state.target_oid.as_deref().ok_or_else(malformed)?,
                    ]
            {
                return Err(AppError::Conflict(
                    "The resulting commit does not match the confirmed merge parents; inspect HEAD"
                        .into(),
                ));
            }
        }
        self.release_integration_owner(
            context,
            session_id,
            if recovery == Recovery::Abort {
                OperationState::Cancelled
            } else {
                OperationState::Succeeded
            },
            match recovery {
                Recovery::Abort => {
                    "Git integration was explicitly aborted; its starting state was restored"
                }
                Recovery::Continue => "Git integration was completed by a confirmed continuation",
                Recovery::Skip => {
                    "The empty cherry-pick was explicitly skipped; HEAD was preserved"
                }
            },
        );
        Ok(())
    }
}

File: apps/desktop/src/core/git/repository/merge.ts
import { badRequest, conflict, malformed, oneLine } from "../support";
import { verifyCherryPick, verifyRevert } from "./cherrypick";
import {
  type Recovery,
  gitIntegration,
  integrationSnapshot,
  ownsIntegration,
  protectAbortPaths,
  releaseIntegrationOwner,
} from "./integration";
import { splitNul } from "./parse";
import { verifyRebase } from "./rebase";
import {
  type Operation,
  RepositoryService,
  type RepositoryContext,
  commandError,
} from "./service";
import { protectLocalPaths, sameHead } from "./stash";
import type { ExpectedState } from "./types";

/**
 * Starting a merge, and resuming any owned integration sequence.
 *
 * A port of `apps/runtime/src/git/repository/integration/merge.rs`. Resume is
 * here rather than beside each starter because the four kinds share one shape:
 * confirm ownership and state, run one Git verb, then prove the repository
 * ended up where the caller was told it would.
 */

export async function startMerge(
  service: RepositoryService,
  context: RepositoryContext,
  targetOid: string,
  message: string,
  expectedToken: string,
  expected: ExpectedState,
  operation: Operation,
): Promise<void> {
  const signal = operation.controller.signal;
  const state = await integrationSnapshot(service, context, signal);
  if (state.kind !== "none") {
    throw conflict("An existing Git operation must be completed first");
  }
  if (!sameHead(state.head, expected) || state.stateToken !== expectedToken) {
    throw conflict(
      "Repository state changed; refresh and confirm the merge again",
    );
  }
  if (
    state.dirty ||
    state.conflicts.length > 0 ||
    state.head.headOid === null ||
    state.head.branch === null
  ) {
    throw conflict(
      "Merge requires a clean worktree and index on a committed local branch",
    );
  }
  if (
    (await service.resolve(context.repository, targetOid, signal)) !== targetOid
  ) {
    throw badRequest("Merge target must be a commit object ID");
  }
  const head = expected.headOid;
  if (head === null) throw malformed();
  const touchedRaw = await service.read(
    context.repository,
    [
      "diff",
      "--no-ext-diff",
      "--no-textconv",
      "--name-only",
      "--no-renames",
      "-z",
      head,
      targetOid,
      "--",
    ],
    signal,
  );
  const touched = new Set(
    splitNul(touchedRaw)
      .filter((path) => path.length > 0)
      .map((path) => path.toString("utf8")),
  );
  await protectLocalPaths(service, context, touched, signal);

  const sessionId = operation.snapshot.id;
  service.integrations.set(context.repository, {
    sessionId,
    kind: "merge",
    mainline: null,
    original: expected,
    targetOid,
    marker: null,
  });
  const command = [
    "-c",
    "rerere.enabled=false",
    "merge",
    "--no-ff",
    "--no-commit",
    "--no-autostash",
    "--no-overwrite-ignore",
    "--no-rerere-autoupdate",
  ];
  if (message !== "") command.push("-m", message);
  command.push("--", targetOid);

  let output;
  let failure: unknown;
  try {
    output = await service.output(
      context.repository,
      command,
      service.commandTimeoutMs,
      signal,
      {
        onMutationStarted: () => {
          operation.mutationStarted = true;
        },
      },
    );
  } catch (error) {
    failure = error;
  }
  // Read with a fresh token after a cancelled child has been reaped: this can
  // bind the marker that is actually left behind without retrying anything.
  const actual = await safely(() => gitIntegration(service, context));
  const actualHead = await safely(() => service.head(context.repository));
  let owns = false;
  if (actual.ok && actualHead.ok) {
    const owner = service.integrations.get(context.repository);
    if (owner !== undefined) {
      if (
        actual.value.kind === "merge" &&
        actual.value.targetOid === targetOid &&
        actual.value.originalHead === expected.headOid &&
        sameHead(actualHead.value, expected)
      ) {
        owner.marker = actual.value.marker;
        owns = ownsIntegration(owner, actual.value, actualHead.value);
      } else if (actual.value.kind === "none") {
        service.integrations.delete(context.repository);
      }
    }
  }
  if (failure !== undefined) throw failure;
  if (!actual.ok) throw actual.error;
  if (!actualHead.ok) throw actualHead.error;
  if (output === undefined) throw malformed();
  if (owns && (output.status === 0 || output.status === 1)) {
    operation.awaitingResolution = true;
    return;
  }
  if (output.status !== 0) throw commandError(output);
  if (actual.value.kind !== "none" || !sameHead(actualHead.value, expected)) {
    throw conflict(
      "Merge ownership or HEAD could not be verified after Git exited; inspect current state",
    );
  }
  const ancestor = await service.output(
    context.repository,
    ["merge-base", "--is-ancestor", targetOid, head],
    15_000,
    signal,
  );
  if (ancestor.status !== 0) {
    throw conflict(
      "Git returned without a merge state or reachable target; inspect the repository",
    );
  }
}

export async function resumeIntegration(
  service: RepositoryService,
  context: RepositoryContext,
  sessionId: string,
  expectedToken: string,
  expected: ExpectedState,
  recovery: Recovery,
  operation: Operation,
): Promise<void> {
  const signal = operation.controller.signal;
  const state = await integrationSnapshot(service, context, signal);
  if (
    !state.owned ||
    state.sessionId !== sessionId ||
    !sameHead(state.head, expected) ||
    state.stateToken !== expectedToken
  ) {
    throw conflict(
      "Integration ownership or repository state changed; refresh before continuing",
    );
  }
  if (recovery === "continue" && !state.canContinue) {
    throw conflict(
      "Resolve all conflicts and stage the saved results before continuing",
    );
  }
  if (recovery === "skip" && !state.canSkip) {
    throw conflict(
      "Only an owned empty cherry-pick or a paused rebase can be skipped; other work must be resolved or explicitly aborted",
    );
  }
  if (recovery !== "continue") {
    if (state.originalHead === null) throw malformed();
    await protectAbortPaths(service, context, state.originalHead, signal);
  }
  const command = recoveryCommand(state.kind, recovery);
  await service.mutate(context, command, operation);
  const after = await gitIntegration(service, context, signal);
  const head = await service.head(context.repository, signal);
  // Abort returns a rebase to its recorded branch, not to the detached head
  // the caller confirmed; every other kind leaves HEAD where it was.
  const restored: ExpectedState =
    state.kind === "rebase"
      ? { headOid: state.originalHead, branch: state.originalBranch }
      : expected;
  if (
    state.kind === "rebase" &&
    recovery !== "abort" &&
    after.kind === "rebase"
  ) {
    // A replay can stop again on the next commit — after a continue and
    // equally after a skip. That is still the same owned sequence.
    const owner = service.integrations.get(context.repository);
    if (owner !== undefined && ownsIntegration(owner, after, head)) {
      operation.awaitingResolution = true;
      return;
    }
  }
  // A finished abort has to land back where the sequence started, and so does
  // a skipped cherry-pick. A skipped rebase is the exception: dropping one
  // replayed commit and finishing the rest is *supposed* to move HEAD.
  const mustRestore =
    recovery === "abort" || (recovery === "skip" && state.kind !== "rebase");
  if (after.kind !== "none" || (mustRestore && !sameHead(head, restored))) {
    throw conflict(
      "Git did not finish the requested recovery; inspect the current state before another action",
    );
  }
  if (recovery === "continue" && state.kind === "cherryPick") {
    if (state.targetOid === null) throw malformed();
    await verifyCherryPick(
      service,
      context,
      state.targetOid,
      expected,
      head,
      signal,
    );
  }
  if (recovery === "continue" && state.kind === "revert") {
    if (state.targetOid === null) throw malformed();
    await verifyRevert(
      service,
      context,
      state.targetOid,
      expected,
      head,
      signal,
    );
  }
  // A rebase that finished is verified whether the last decision was a
  // continue or a skip: both end with the branch back on itself and the
  // confirmed target reachable.
  if (recovery !== "abort" && state.kind === "rebase") {
    await verifyRebase(service, context, state, head, signal);
  }
  if (recovery === "continue" && state.kind === "merge") {
    const commit = head.headOid;
    if (commit === null) throw malformed();
    const parentLine = oneLine(
      await service.read(
        context.repository,
        ["rev-list", "--parents", "-n", "1", commit, "--"],
        signal,
      ),
    );
    const parents = parentLine
      .split(/\s+/)
      .slice(1)
      .filter((v) => v !== "");
    if (expected.headOid === null || state.targetOid === null)
      throw malformed();
    if (
      head.branch !== expected.branch ||
      parents.length !== 2 ||
      parents[0] !== expected.headOid ||
      parents[1] !== state.targetOid
    ) {
      throw conflict(
        "The resulting commit does not match the confirmed merge parents; inspect HEAD",
      );
    }
  }
  releaseIntegrationOwner(
    service,
    context,
    sessionId,
    recovery === "abort" ? "cancelled" : "succeeded",
    recovery === "abort"
      ? "Git integration was explicitly aborted; its starting state was restored"
      : recovery === "continue"
        ? "Git integration was completed by a confirmed continuation"
        : state.kind === "rebase"
          ? "The replayed commit the rebase stopped on was explicitly skipped"
          : "The empty cherry-pick was explicitly skipped; HEAD was preserved",
  );
}

function recoveryCommand(kind: string, recovery: Recovery): string[] {
  if (kind === "merge" && recovery === "abort") return ["merge", "--abort"];
  if (kind === "merge" && recovery === "continue") {
    return ["-c", "core.editor=:", "commit", "--no-edit"];
  }
  if (kind === "cherryPick") {
    return [
      "-c",
      "core.editor=:",
      "cherry-pick",
      recovery === "continue"
        ? "--continue"
        : recovery === "abort"
          ? "--abort"
          : "--skip",
    ];
  }
  // Skip belongs to an empty cherry-pick only; dropping a revert silently
  // would leave the change it was meant to undo in place.
  if (kind === "revert" && recovery !== "skip") {
    return [
      "-c",
      "core.editor=:",
      "revert",
      recovery === "continue" ? "--continue" : "--abort",
    ];
  }
  if (kind === "rebase") {
    return [
      "-c",
      "core.editor=:",
      "-c",
      "rerere.enabled=false",
      "rebase",
      recovery === "continue"
        ? "--continue"
        : recovery === "abort"
          ? "--abort"
          : "--skip",
    ];
  }
  throw conflict("This Git operation is not managed by the current service");
}

/** Runs `work`, reporting the failure instead of throwing it. */
export async function safely<T>(
  work: () => Promise<T>,
): Promise<{ ok: true; value: T } | { ok: false; error: unknown }> {
  try {
    return { ok: true, value: await work() };
  } catch (error) {
    return { ok: false, error };
  }
}
